lib/internal/streams/end-of-stream.js

```diff
@@ -63,22 +63,47 @@ function eos(stream, opts, callback) {
 
   const wState = stream._writableState;
   const rState = stream._readableState;
+  const state = wState || rState;
 
   const onlegacyfinish = () => {
     if (!stream.writable) onfinish();
   };
 
+  // TODO (ronag): Improve soft detection to include core modules and
+  // common ecosystem modules that do properly emit 'close' but fail
+  // this generic check.
+  const willEmitClose = (
+    state &&
+    state.autoDestroy &&
+    state.emitClose &&
+    state.closed === false
+  );
+
   let writableFinished = stream.writableFinished ||
     (wState && wState.finished);
   const onfinish = () => {
     writableFinished = true;
+    if (willEmitClose && (!stream.readable || readable)) {
+      // Duplex won't autoDestroy if `stream.readable` was
+      // explicitly set to false. Bug?
```

Comment:
The question mark would make people wonder of why this change was made. Maybe add in a TODO to investigate?